edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
+
+        base_counts = self.counts / effective_lengths
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
+        if log:
+            self.counts[self.counts == 0] = prior_count
+            self.counts = np.log(self.counts)
 
```

Comment:
IMHO you should just return self here to allow for chaining of functions in the future. Same for other norm. methods.

Like this theoretical:

```python
DGEList().from_something().rpkm().save_to_disk()
```

Reply:
I think this is counter intuitive (as I suggested in the PR):
Either we have an imperative function rpkm (read "rpkm-normalize yourself!", which means that the object should transform/normalise itself) or we have a pure rpkm (read "return rpkm normalized counts" which means that the object doesn't modify itself). Or we have a Javascript like chaining function as you suggest.

Intuitively to me, the name rpkm suggests the second, I don't see a reason to adopt the method chaining pattern from Javascript here as it does not seem intuitively to me.

Reply:
That's OK. Mutating `self` and then returning `self` lends itself to chaining. 

IMHO if you are mutating self and returning `None` then returning `self` is free and let's the user of the API chain if they want. Not a dealbreaker for me but let's for pretty powerful queries like `object.filter().drop().normalize()`.

Reply:
However, I agree with the argument that a mutation of an object makes it so previous states are lost (at a savings of memory) and that may be an annoying behavior for scientific exploration.

Since these object can be quite large (thousands of genes and/or samples) I think mutating the original is fair but if a user wants to keep a previous state around, an explicit cast to a `.copy()` method should be made and the expectation would be a memory copy (deep copy).

I guess this would be a call for:

```python
class DGEList(object):
    
    def copy(self) -> 'DGEList':
        ...
        DGEList(self.counts.copy(), ...)
```

Reply:
@apfejes what do you think?

Reply:
IMO either we should employ chaining as a paradigm for the whole library and implement it for every function (if possible) or not at all.
The copy function is necessary anyways. I think such a function is related to read/write_npz_file (as those functions access/store all data that would need to be copied.